### PR TITLE
Add test for input

### PR DESCRIPTION
**Pull request by allevo**
# Objective

Create some e2e tests. Place a starting point for discussion

## Solution

The purpose is to test (just for now) the events and how bevy handles them.
The solution is not OK:
- a lot of copy and paste
- the macro could be generalized and put in a dedicated file (which one? I'm open)

### CODE REVIEW

File: tests/input.rs [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+#[test]
+fn test_input_received_character_single_button() {
+    let mut app = App::new();
+
+    app.add_plugin(CorePlugin::default());
+    app.add_plugin(WindowPlugin::default());
+    app.add_plugin(AssetPlugin::default());
+    app.add_plugin(RenderPlugin::default());
+    app.add_plugin(CorePipelinePlugin::default());
+    app.add_plugin(PbrPlugin::default());
```
Comment (mockersf) [inline tests/input.rs:64]: Why are you adding all those plugins? I think most are not needed for those test.

That would remove the need to install additional dependencies for tests too

File: tests/input.rs [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+
+macro_rules! send_event {
+    ($app: ident, $t: ty, $event: expr) => {
+        let world = &mut $app.world;
+        let mut sender = world.get_resource_mut::<Events<$t>>().unwrap();
+        sender.send($event);
+    };
+}
+
+#[cfg_attr(target_os = "linux")]
```
Comment (mockersf) [inline tests/input.rs:54]: this shouldn't be needed

File: tests/input.rs [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+    assert_eq!(entities.len(), 0);
+
+    send_event!(
+        app,
+        ReceivedCharacter,
+        ReceivedCharacter {
+            id: WindowId::primary(),
+            char: 'a',
+        }
+    );
```
Comment (mockersf) [inline tests/input.rs:80]: I don't think this macro is needed

File: tests/input.rs [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+
+    app.update();
+
+    let chars: HashSet<_> = get_entities!(app, (Entity, &CharComponent))
+        .into_iter()
+        .map(|(_, c)| c.0)
+        .collect();
```
Comment (mockersf) [inline tests/input.rs:87]: Same here, not sure this macro is needed

**Review by mockersf — CHANGES_REQUESTED** [review]

File: tests/input.rs (hunk 1 above)
Comment (allevo) [reply tests/input.rs:64]: Yes you are right, I copied from a tests that use sprites, materials and so on...
I'ill removed

File: tests/input.rs (hunk 3 above)
Comment (allevo) [reply tests/input.rs:80]: the macros simplify the readability of the test.
Avoids also the copy-and-paste.
Should I remove it ?

File: tests/input.rs [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+macro_rules! get_entities {
+    ($app: ident, $query: tt) => {{
+        let world = &mut $app.world;
+        let mut query = world.query::<$query>();
+        query.iter(world).collect::<Vec<_>>()
+    }};
+}
+
+macro_rules! send_event {
+    ($app: ident, $t: ty, $event: expr) => {
```
Comment (Sheepyhead) [inline tests/input.rs:42]: Is it possible to get the type from the input without having to repeat the type as argument?

**Review by Sheepyhead — COMMENTED** [review]
As a general comment I find the intent of the tests somewhat hard to parse by reading the code. A few comments might clarify what exactly a test is testing

File: tests/input.rs (hunk 3 above)
Comment (alice-i-cecile) [reply tests/input.rs:67]: I would prefer to avoid macros here.

File: tests/input.rs [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+            char: 'B',
+        }
+    );
+
+    app.update();
+
+    let entities: Vec<_> = get_entities!(app, Entity);
+    assert_eq!(entities.len(), 2);
+
+    (0..10).for_each(|_| app.update());
```
[the remark this answers is not in the document]
Comment (allevo) [reply tests/input.rs:253]: added an assertion after, sorry for the forgot lines

**Comment by allevo** [comment]
> As a general comment I find the intent of the tests somewhat hard to parse by reading the code. A few comments might clarify what exactly a test is testing

For me the tests are pretty clear but I wrote them, so probably for me it is easier to read! 😄 

The aims for those tests are:
- explain how to test a system function that use an `EventReader`
- check if the event system works properly with the App lifecycle
  - if an event is sent, the system receives it
  - if multiple events are sent, the system receives all of them
  - the event, if consumed by the system, is not available anymore

At least the above lines are my intentions.
If you are ok with the tests, I really would like to be clearer with comments or other stuff

Let me know.

File: tests/input.rs [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+    input::InputPlugin,
+    input::{keyboard::KeyboardInput, ButtonState},
+    prelude::App,
+    prelude::{EventReader, KeyCode},
+    window::{ReceivedCharacter, WindowId, WindowPlugin},
+};
+
+#[test]
+fn test_input_received_character() {
+    let inputs = Arc::new(Mutex::new(VecDeque::from(['a', 'B'])));
```
Comment (alice-i-cecile) [inline tests/input.rs:17]: This seems extremely elaborate. Why doesn't a simple vec work?

File: tests/input.rs [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+    let inputs2 = inputs.clone();
+    let system = move |mut chat_input_events: EventReader<ReceivedCharacter>| {
```
Comment (alice-i-cecile) [inline tests/input.rs:19]: The closure system is clever but makes this harder to read. I would just use a standard `fn` system.

File: tests/input.rs [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    // Multiple events can be sent multiple time per `update`.
+    // The iterator inside the system function returns 2 events
+    app.update();
+    assert_eq!(expected_keyboard_inputs.lock().unwrap().len(), 0);
+    assert_eq!(expected_received_chars.lock().unwrap().len(), 0);
+}
+
+fn send_event<E: Event>(app: &mut App, event: E) {
+    let world = &mut app.world;
+    let mut sender = world.get_resource_mut::<Events<E>>().unwrap();
```
Comment (alice-i-cecile) [inline tests/input.rs:224]:
```suggestion
    let mut sender = world.resource_mut::<Events<E>>();
```

File: tests/input.rs [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+    );
+
+    // Multiple events can be sent multiple time per `update`.
+    // The iterator inside the system function returns 2 events
+    app.update();
+    assert_eq!(expected_keyboard_inputs.lock().unwrap().len(), 0);
+    assert_eq!(expected_received_chars.lock().unwrap().len(), 0);
+}
+
+fn send_event<E: Event>(app: &mut App, event: E) {
```
Comment (alice-i-cecile) [inline tests/input.rs:222]: I think this should just be a method on `App` and `World`. Could you split this out into its own PR?

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Left some feedback. I think we can make this testing strategy much simpler :)

@allevo can you mark comments that you've completed as resolved?

**Comment by richchurcher** [comment]
Closing as part of backlog cleanup: long period of inactivity, likely significant bit rot by now.
